src/mlpack/methods/ne/cne.hpp

```diff
@@ -0,0 +1,186 @@
+ /**
+ * @file cne.hpp
+ * @author Bang Liu
+ *
+ * Definition of CNE class.
+ */
+#ifndef MLPACK_METHODS_NE_CNE_HPP
+#define MLPACK_METHODS_NE_CNE_HPP
+
+#include <cstddef>
+#include <cstdio>
+
+#include <mlpack/core.hpp>
+
+#include "link_gene.hpp"
+#include "neuron_gene.hpp"
+#include "genome.hpp"
+#include "species.hpp"
+#include "tasks.hpp"
+#include "parameters.hpp"
+
+namespace mlpack {
+namespace ne {
+
+/**
+ * This class implements Conventional Neuro-evolution (CNE): weight
+ * evolution on topologically fixed neural networks.
+ */
+template<typename TaskType>
+class CNE {
+ public:
+  // Parametric constructor.
+  CNE(TaskType task, Genome& seedGenome, Parameters& params) {
+    aTask = task;
+    aSeedGenome = seedGenome;
+    aSpeciesSize = params.aSpeciesSize;
+    aMaxGeneration = params.aMaxGeneration;
+    aMutateRate = params.aMutateRate;
+    aMutateSize = params.aMutateSize;
+    aElitePercentage = params.aElitePercentage;
+  }
+
+  // Destructor.
+  ~CNE() {}
+
+  // Soft mutation: add a random value chosen from
+  // initialization prob distribution with probability p.
+  // TODO: here we use uniform distribution. Can we use exponential distribution?
+  static void MutateWeightsBiased(Genome& genome, double mutateProb, double mutateSize) {
+    for (int i=0; i<genome.aLinkGenes.size(); ++i) {
+      double p = mlpack::math::Random();  // rand 0~1
+      if (p < mutateProb) {
+        double deltaW = mlpack::math::RandNormal(0, mutateSize);
+        double oldW = genome.aLinkGenes[i].Weight();
+        genome.aLinkGenes[i].Weight(oldW + deltaW);
+      }
+    }
+  }
+
+  // Hard mutation: replace with a random value chosen from
+  // initialization prob distribution with probability p.
+  // TODO: here we use uniform distribution. Can we use exponential distribution?
+  static void MutateWeightsUnbiased(Genome& genome, double mutateProb, double mutateSize) {
```

Comment:
Do you think, we could use a boolean to distinguish between soft and hard mutation. In this case we could reuse the basic function.